### PR TITLE
Add typescript

### PR DESCRIPTION
**Pull request by ayinloya**
Add types to sdk
Generate js from typescript

Resolves https://github.com/smileidentity/smile-identity-core-js/issues/23

### CODE REVIEW

File: index.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import { IDApi } from './src/id-api';
+import Signature from './src/signature';
+import { Utilities } from './src/utilities';
+import { WebApi } from './src/web-api';
+import { JOB_TYPE, IMAGE_TYPE } from './src/constants';
+
+if ((typeof process === 'undefined' && typeof (process as NodeJS.Process).versions.node === 'undefined') || typeof window !== 'undefined') {
+  console.error('This is a server-side library meant for a node.js (or compatible) runtime, and is not meant to work in the browser.');
```
Comment (cmrfrd) [inline index.ts:8]: Maybe this should be a `warning` opposed to an `error`.

File: src/constants.ts [hunk 2]
```diff
@@ -59,9 +59,7 @@ const JOB_TYPE = {
   COMPARE_USER_INFO: 9,
 };
 
-const sidServerMapping = {
+export const sidServerMapping : { [k:string | number]: string } = {
```
Comment (cmrfrd) [inline src/constants.ts:62]: Is there a way we can make this type strictly `number` or strictly `string`?

File: src/shared.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
+interface OptionsParam {
```
Comment (cmrfrd) [inline src/shared.ts:1]: Can we follow the same export pattern as in [src/helpers.ts](https://github.com/smileidentity/smile-identity-core-js/pull/58/files#diff-ca5696b367d474e785317cb7a0d9853fef5729387ab8d0fab4c46268c03aae99)

File: tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,106 @@
+{
+  "include": [
+    "./index.ts",
+    "src/**/*"],
+  "compilerOptions": {
+    /* Visit https://aka.ms/tsconfig to read more about this file */
+
+    /* Projects */
```
Comment (cmrfrd) [inline tsconfig.json:8]: Can we remove all the extra commented out tsconfig options

File: package.json [hunk 5]
```diff
@@ -2,10 +2,11 @@
...
   "version": "2.0.0",
   "description": "The official Smile Identity Node SDK",
-  "main": "index.js",
+  "main": "dist/index.js",
   "scripts": {
-    "lint": "eslint $(git ls-files '*.js')",
-    "start": "node index.js",
+    "build": "tsc",
+    "lint": "eslint $(git ls-files '*.js' '*.ts')",
+    "start": "node dist/index.js",
```
Comment (mldangelo) [inline package.json:9]: This may not exist. You should take another pass on setting up scripts.

File: package.json [hunk 6]
```diff
@@ -2,10 +2,11 @@
...
     "test": "jest --coverage"
   },
   "author": "Chris DeLauder <chris@smileidentity.com>",
```
Comment (mldangelo) [inline package.json:12]: @ayinloya you should add yourself to this

File: package.json [hunk 7]
```diff
@@ -2,10 +2,11 @@
   "name": "smile-identity-core",
   "version": "2.0.0",
   "description": "The official Smile Identity Node SDK",
-  "main": "index.js",
+  "main": "dist/index.js",
```
Comment (mldangelo) [inline package.json:5]: should we add a "typings" definition next to this?

File: package.json [hunk 8]
```diff
@@ -2,10 +2,11 @@
   "name": "smile-identity-core",
   "version": "2.0.0",
```
Comment (mldangelo) [inline package.json:3]:
```suggestion
  "version": "3.0.0-pre",
```

File: examples/server.js [hunk 9]
```diff
@@ -1,10 +1,8 @@
...
-const express = require('express');
-const { json } = require('express');
-const express_static = require('express').static;
-const UUID = require('uuid').v4;
-const { config } = require('dotenv');
-const { WebApi } = require('smile-identity-core');
+import { express, static as express_static, json } from 'express';
+import { v4 as UUID } from 'uuid';
+import { config } from 'dotenv';
+import { WebApi } from 'smile-identity-core';
```
Comment (mldangelo) [inline examples/server.js:5]: this change will break node 12 compatibility

File: examples/package.json [hunk 10]
```diff
@@ -5,7 +5,9 @@
   "main": "server.js",
   "scripts": {},
   "dependencies": {
-    "smile-identity-core": "latest",
+    "dotenv": "^16.0.3",
+    "express": "^4.18.2",
+    "smile-identity-core": "file:../",
```
Comment (mldangelo) [inline examples/package.json:10]: nice change

File: .github/workflows/npm-publish.yml [hunk 11]
```diff
@@ -14,6 +14,7 @@ jobs:
           node-version-file: '.nvmrc'
           cache: 'npm'
       - run: npm ci
+      - run: npm run build
```
Comment (mldangelo) [inline .github/workflows/npm-publish.yml:17]: Do you need to do this for the tests, or do you do this as a test? 

You should also do this as a publish step, or implement it as a prepublish step.

File: src/constants.ts (hunk 2 above)
Comment (ayinloya) [reply src/constants.ts:62]: sure, will have to make the changes to see

File: package.json (hunk 7 above)
Comment (ayinloya) [reply package.json:5]: I'm not sure how to add the typings when there are multiple declarations, will have to look it up
maybe @cmrfrd can assist with that

File: examples/server.js (hunk 9 above)
Comment (ayinloya) [reply examples/server.js:5]: okay, will revert to require

File: .github/workflows/npm-publish.yml (hunk 11 above)
Comment (ayinloya) [reply .github/workflows/npm-publish.yml:17]: This is done as a test to be sure the compilation works

Comment (ayinloya) [reply .github/workflows/npm-publish.yml]: Changed this, now doing type check in lint script and building js and d.ts in republish step

File: package.json (hunk 7 above)
Comment (cmrfrd) [reply package.json:5]: Adding multiple declarations should be straight forward, it's very much similar to c/c++ headers.

https://www.typescriptlang.org/docs/handbook/declaration-files/templates/module-d-ts.html

File: examples/package.json (hunk 10 above)
Comment (ayinloya) [reply examples/package.json:10]: thanks :)

File: package.json (hunk 5 above)
Comment (ayinloya) [reply package.json:10]: done, this is now generated when npm install is run

**Review by mldangelo — APPROVED** [review]
This looks good to me. Please feel free to merge when you are ready.

**Pull request merged**